pkg/cmd/image/list.go

```diff
@@ -188,50 +189,126 @@ type imagePrinter struct {
+		blobSize += layerDescriptor.Size
+	}
+
+	// Get the platform
+	plt := platforms.Normalize(ocispec.Platform{OS: config.OS, Architecture: config.Architecture, Variant: config.Variant})
+
+	// Get the filesystem size for all layers
+	chainID := identity.ChainID(config.RootFS.DiffIDs).String()
+	size := int64(0)
+	if _, actualSize, err := imgutil.ResourceUsage(ctx, snapshotter, chainID); err == nil {
+		size = actualSize.Size
+	}
+
+	return &Image{
+		blobSize: blobSize,
+		size:     size,
+		platform: plt,
+		config:   &manifest.Config,
+	}, nil
 }
 
-func makePlatformKey(platform v1.Platform) string {
-	if platform.OS == "" {
-		return "unknown"
+func ReadIndex(ctx context.Context, provider containerdutil.Provider, snapshotter snapshots.Snapshotter, desc ocispec.Descriptor) (map[string]*Image, error) {
+	descs := map[string]*Image{}
+
+	// Read the index
+	indexData, err := ReadBlob(ctx, provider, desc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal as Index
+	var index ocispec.Index
+	if err := json.Unmarshal(indexData, &index); err != nil {
+		return nil, err
 	}
 
-	return path.Join(platform.OS, platform.Architecture, platform.OSVersion, platform.Variant)
+	// Iterate over manifest descriptors and read them all
+	for _, manifestDescriptor := range index.Manifests {
+		manifest, err := ReadManifest(ctx, provider, snapshotter, manifestDescriptor)
+		if err != nil {
+			continue
+		}
+		descs[platforms.FormatAll(manifest.platform)] = manifest
+	}
+	return descs, err
 }
 
-func (x *imagePrinter) printImageSinglePlatform(ctx context.Context, img images.Image, ociPlatform v1.Platform) error {
-	platMC := platforms.OnlyStrict(ociPlatform)
-	if avail, _, _, _, availErr := images.Check(ctx, x.contentStore, img.Target, platMC); !avail {
-		log.G(ctx).WithError(availErr).Debugf("skipping printing image %q for platform %q", img.Name, platforms.Format(ociPlatform))
-		return nil
+func Read(ctx context.Context, provider containerdutil.Provider, snapshotter snapshots.Snapshotter, desc ocispec.Descriptor) (map[string]*Image, error) {
+	switch desc.MediaType {
+	case images.MediaTypeDockerSchema2Manifest, ocispec.MediaTypeImageManifest:
+		manifest, err := ReadManifest(ctx, provider, snapshotter, desc)
+		if err != nil {
+			return nil, err
+		}
+		descs := map[string]*Image{}
+		descs[platforms.FormatAll(manifest.platform)] = manifest
+		return descs, nil
+	case images.MediaTypeDockerSchema2ManifestList, ocispec.MediaTypeImageIndex:
+		return ReadIndex(ctx, provider, snapshotter, desc)
 	}
+	return nil, fmt.Errorf("nothing")
+}
 
-	image := containerd.NewImageWithPlatform(x.client, img, platMC)
-	desc, err := image.Config(ctx)
-	if err != nil {
-		log.G(ctx).WithError(err).Warnf("failed to get config of image %q for platform %q", img.Name, platforms.Format(ociPlatform))
+func (x *imagePrinter) printImage(ctx context.Context, img images.Image) error {
+	candidateImages, _ := Read(ctx, x.contentStore, x.snapshotter, img.Target)
```

Comment:
Error shouldn't be ignored

Reply:
Yep, not sure why it escaped me.
Thanks a lot for catching it.
Next push will fix it.